File: LeetCode/iter1/c++/611.valid-triangle-number.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:

	int triangleNumber(vector<int>& nums) {
		sort(nums.begin(), nums.end());

		int cnt = 0;
		int l = 0, r = 0;
		for (int i = nums.size() - 1; i >= 2; i--) {
			l = 0;
			r = i - 1;
			while (l < r) {
				if (nums[l] + nums[r] > nums[i]) {
					cnt += r - l; //l+r>nums[i],说明l+1, l+2与r, nums[i]可以构成有效三角形
					r--;
				}
				else {
					l++;
				}
			}
		}

		return cnt;
	}
};

int main(void)
{
	Solution solution;
	vector<int> nums;
	nums.push_back(2);
	nums.push_back(2);
	nums.push_back(3);
	nums.push_back(4);

	cout << solution.triangleNumber(nums) << endl;

	return 0;
}

File: LeetCode/iter1/java/611.valid-triangle-number.java
package org.ironman.leetcode;

import java.util.Arrays;

public class Main {

    public static int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int cnt = 0;
        int l, r;
        for (int i = nums.length - 1; i >= 2; i--) {
            l = 0;
            r = i - 1;
            while (l < r) {
                if (nums[l] + nums[r] > nums[i]) {
                    cnt += r - l; //if nums[l]+nums[r] > nums[i], then nums[l+1] or nums[l+2] + nums[r] > nums[i]
                    r--;
                } else {
                    l++;
                }
            }
        }
        return cnt;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{2, 2, 3, 4};
        System.out.println(triangleNumber(nums));
    }
}
